File: Striver/Day 2 : Array 2/Merge Sorted Array.cpp
//Question: 

//Solution : Method 1 : O(n) extra space O(m+n) time
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int p1=0,p2=0;
        vector <int> temp;
        while(p1 < m and p2 < n){
            if (nums2[p2] < nums1[p1]){
                temp.push_back(nums2[p2++]);
            }
            else temp.push_back(nums1[p1++]);
        }
        while(p1<m){
            temp.push_back(nums1[p1++]);
        }
        while(p2<n){
            temp.push_back(nums2[p2++]);
        }
        nums1 = temp;
    }
};

//Method 2: O(1) extra space O(n*m)
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int p1=0;
        if (nums2.empty()) return;
        while(p1<m){
            if (nums1[p1] <= nums2[0]){
                p1++;
                continue;
            }
            else{
                swap(nums1[p1],nums2[0]);
                sort(nums2.begin(),nums2.end());  //can be done in linear too
            }
        }
        for (int i=0; i<n; i++){
            nums1[m+i] = nums2[i];
        }
    }
};

File: Striver/Day 2 : Array 2/Merge Sorted Array.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        
        
       for(int i=0 ,j=0 ; i<nums1.length ;i++)
       {
           if(nums1[i] == 0)
           {
                if(j==nums2.length)
                   break;
               
               int temp = nums2[j];
               nums2[j] = nums1[i];
               nums1[i] = temp;
               j++;
               
           }
       }
        
      Arrays.sort(nums1);  
        
    }
}
